### PR TITLE
UCP: Implementing Reduce Knomial

### PR DESCRIPTION
**Pull request by shimmybalsam**
## What
Implemendting Redice Knomial algorithm

## Why ?
To have as part of UCC library.

## How ?
Opposite of bcast KN, allocating scratch buffers for each step reduction.

### CODE REVIEW

**Comment by artemry-nv** [comment]
bot:retest

**Comment by artemry-nv** [comment]
bot:retest

File: test/gtest/core/test_reduce.cc [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            }
+        }
+    }
+    void data_fini(UccCollCtxVec ctxs) {
+    	for (auto r = 0; r < ctxs.size(); r++) {
+            ucc_coll_args_t* coll = ctxs[r]->args;
+            if (r == root) {
+                UCC_CHECK(ucc_mc_free(ctxs[r]->dst_mc_header));
+            }
+            if (coll->src.info.buffer) { /* no inplace */
```
Comment (lappazos) [inline test/gtest/core/test_reduce.cc:90]: why not the same condition as alloc?

File: tools/perf/ucc_pt_coll_reduce.cc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#include <utils/ucc_coll_utils.h>
+
+ucc_pt_coll_reduce::ucc_pt_coll_reduce(ucc_datatype_t dt, ucc_memory_type mt,
+                        ucc_reduction_op_t op, bool is_inplace, bool is_root)
+{
+    has_inplace_   = true;
+    has_reduction_ = true;
+    has_range_     = true;
+    has_bw_        = true;
+    is_root_       = is_root;
```
Comment (lappazos) [inline tools/perf/ucc_pt_coll_reduce.cc:14]: why not just do is_root_ = rank==0, and then you won't need to change all perf files to add is_root

File: tools/perf/ucc_pt_coll_reduce.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    if (is_root_) {
+        ucc_mc_free(dst_header);
+    }
+}
+
+float ucc_pt_coll_reduce::get_bw(float time_ms, int grsize,
+                                    ucc_coll_args_t args)
+{
+    float S = args.src.info.count * ucc_dt_size(args.src.info.datatype);
```
Comment (lappazos) [inline tools/perf/ucc_pt_coll_reduce.cc:73]: @bureddy why is this the way to calculate bw? 
for root itself, he will get way more than 1 vector along the algorithm (he will get ~ vector * radix* dist), and in general the amount of data in the network is larger

**Review by lappazos — CHANGES_REQUESTED** [review]

File: src/components/tl/ucp/reduce/reduce_knomial.c [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    UCC_REDUCE_KN_GOTO_PHASE(task->reduce_kn.phase);
+
+//UCC_REDUCE_KN_PHASE_LOOP:
+    while (task->bcast_kn.dist <= task->reduce_kn.max_dist) {
+        if (vrank % task->bcast_kn.dist == 0) {
+            pos = (vrank / task->bcast_kn.dist) % radix;
+            if (pos == 0) {
+            	scratch_offset = received_vectors;
+            	task->reduce_kn.children_per_cycle = 0;
+                for (i = radix - 1; i >= 1; i--) {
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/reduce/reduce_knomial.c:63]: still needs to be smaller than radix per iteration but will changed from 1 to radix -1 with break.

File: src/components/tl/ucp/reduce/reduce.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+/* UCC_REDUCE_KN_CHECK_PHASE(UCC_REDUCE_KN_PHASE_LOOP);               \ */
+#define UCC_REDUCE_KN_GOTO_PHASE(_phase)                                       \
+    do {                                                                       \
+        switch (_phase) {                                                      \
+            UCC_REDUCE_KN_CHECK_PHASE(UCC_REDUCE_KN_PHASE_MULTI);              \
+            UCC_REDUCE_KN_CHECK_PHASE(UCC_REDUCE_KN_PHASE_PROGRESS);           \
+        case UCC_REDUCE_KN_PHASE_INIT:                                         \
+            break;                                                             \
+        };                                                                     \
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/reduce/reduce.h:33]: I think its ok

File: src/components/tl/ucp/reduce/reduce_knomial.c [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "reduce.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+#include "utils/ucc_math.h"
+
+#define CALC_DIST(_size, _radix, _dist)                                        \
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/reduce/reduce_knomial.c:14]: moved to tl_ucp_coll.h

File: test/gtest/core/test_reduce.cc (hunk 1 above)
Comment (shimmybalsam) [reply test/gtest/core/test_reduce.cc:90]: You're right, I now changed it to be aligned with alloc.

File: src/components/tl/ucp/reduce/reduce.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+#ifndef REDUCE_H_
+#define REDUCE_H_
+#include "../tl_ucp.h"
+#include "../tl_ucp_coll.h"
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/reduce/reduce.h:9]: It's ok

File: tools/perf/ucc_pt_coll_reduce.cc (hunk 3 above)
Comment (shimmybalsam) [reply tools/perf/ucc_pt_coll_reduce.cc:73]: This isn't implementation specific calculation. It is an estimation of best band width cascanario (for example ring implementation) using latency of our implementation (which is better latency) in order to analyze real bw of our implementation.

File: src/components/tl/ucp/reduce/reduce.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+enum {
+    UCC_REDUCE_KN_PHASE_INIT,
+    UCC_REDUCE_KN_PHASE_PROGRESS, /* checks progress */
+    UCC_REDUCE_KN_PHASE_MULTI     /* reduce multi after recv from children in current step */
+};
+
+#define UCC_REDUCE_KN_CHECK_PHASE(_p)                                          \
+    case _p:                                                                   \
+        goto _p;                                                               \
+        break;
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/reduce/reduce.h:23]: removed

File: src/components/tl/ucp/reduce/reduce.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+/* UCC_REDUCE_KN_CHECK_PHASE(UCC_REDUCE_KN_PHASE_LOOP);               \ */
+#define UCC_REDUCE_KN_GOTO_PHASE(_phase)                                       \
+    do {                                                                       \
+        switch (_phase) {                                                      \
+            UCC_REDUCE_KN_CHECK_PHASE(UCC_REDUCE_KN_PHASE_MULTI);              \
+            UCC_REDUCE_KN_CHECK_PHASE(UCC_REDUCE_KN_PHASE_PROGRESS);           \
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/reduce/reduce.h]: It is good to keep for future adjustments.

File: tools/perf/ucc_pt_coll_alltoall.cc [hunk 10]
```diff
@@ -29,9 +30,10 @@ ucc_pt_coll_alltoall::ucc_pt_coll_alltoall(int size, ucc_datatype_t dt,
 ucc_status_t ucc_pt_coll_alltoall::init_coll_args(size_t single_rank_count,
                                                   ucc_coll_args_t &args)
 {
-    size_t       dt_size = ucc_dt_size(coll_args.src.info.datatype);
-    size_t       size    = comm_size * single_rank_count * dt_size;
-    ucc_status_t st      = UCC_OK;
+	int          comm_size = comm->get_size();
```
Comment (lappazos) [inline tools/perf/ucc_pt_coll_alltoall.cc:33]: aligment

File: tools/perf/ucc_pt_coll_alltoallv.cc [hunk 11]
```diff
@@ -29,9 +30,10 @@ ucc_pt_coll_alltoallv::ucc_pt_coll_alltoallv(int size, ucc_datatype_t dt,
 ucc_status_t ucc_pt_coll_alltoallv::init_coll_args(size_t count,
                                                    ucc_coll_args_t &args)
 {
-    size_t       dt_size = ucc_dt_size(coll_args.src.info_v.datatype);
-    size_t       size    = comm_size * count * dt_size;
-    ucc_status_t st      = UCC_OK;
+	int          comm_size = comm->get_size();
```
Comment (lappazos) [inline tools/perf/ucc_pt_coll_alltoallv.cc:33]: aligment

**Review by lappazos — APPROVED** [review]

File: src/components/tl/ucp/tl_ucp.c [hunk 12]
```diff
@@ -65,6 +65,10 @@ static ucc_config_field_t ucc_tl_ucp_lib_config_table[] = {
      ucc_offsetof(ucc_tl_ucp_lib_config_t, bcast_kn_radix),
      UCC_CONFIG_TYPE_UINT},
 
+    {"REDUCE_KN_RADIX", "4", "Radix of the recursive-knomial reduce algorithm",
```
Comment (vspetrov) [inline src/components/tl/ucp/tl_ucp.c:68]: recursive-knomial is wrong name: it is knomial-tree, there is not recursion

File: src/components/tl/ucp/tl_ucp_coll.h [hunk 13]
```diff
@@ -17,6 +17,14 @@
 extern const char
     *ucc_tl_ucp_default_alg_select_str[UCC_TL_UCP_N_DEFAULT_ALG_SELECT_STR];
 
+#define CALC_DIST(_size, _radix, _dist)                                        \
```
Comment (vspetrov) [inline src/components/tl/ucp/tl_ucp_coll.h:20]: Lets call it CALC_KN_TREE_DIST

File: test/gtest/core/test_reduce.cc [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+                typename T::type * ptr;
+                ptr = (typename T::type *)ctxs[r]->init_buf;
+                /* need to limit the init value so that "prod" operation
+                   would not grow too large. We have teams up to 16 procs
+                   in gtest, this would result in prod ~2**48 */
+                ptr[i] = (typename T::type)((i + r + 1) % 8);
+            }
+
+            if (r == root) {
+			    coll->dst.info.mem_type = mem_type;
```
Comment (vspetrov) [inline test/gtest/core/test_reduce.cc:56]: alignment (looks 8 spaces or tab)

File: test/gtest/core/test_reduce.cc [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+                this->data_fini(ctxs);                                         \
+            }                                                                  \
+        }                                                                      \
+    }
+
+TYPED_TEST(test_reduce, single_host) {
+    TEST_DECLARE(UCC_MEMORY_TYPE_HOST, TEST_NO_INPLACE, 1);
+}
+
+TYPED_TEST(test_reduce, single_host_persistent)
```
Comment (vspetrov) [inline test/gtest/core/test_reduce.cc:172]: please keep "{" on the same line - make it uniform in all TYPED_TEST declarations

**Review by vspetrov — APPROVED** [review]
Minor stylistic comments only. Plz add some basic performance evaluation. For example: osu_reduce 2, 8 nodes with different ppn. We can compare vs ompi/coll/tuned.

File: src/components/tl/ucp/reduce/reduce_knomial.c [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+            pos = (vrank / task->bcast_kn.dist) % radix;
+            if (pos == 0) {
+                scratch_offset = received_vectors;
+                task->reduce_kn.children_per_cycle = 0;
+                for (i = 1; i < radix; i++) {
+                    vpeer = vrank + i * task->bcast_kn.dist;
+                    if (vpeer >= team_size) {
+                    	break;
+                    } else {
+                        task->reduce_kn.children_per_cycle += 1;
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/reduce/reduce_knomial.c:62]: After discussion on call with @vspetrov, decided to keep as is with children_per_cycle, as task->recv_posted does not re-initiate to zero per dist.

File: src/components/tl/ucp/reduce/reduce_knomial.c [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+ucc_status_t ucc_tl_ucp_reduce_knomial_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      =
+        ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_coll_args_t   *args      = &coll_task->args;
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    uint32_t           radix     = task->reduce_kn.radix;
+    ucc_rank_t         myrank    = team->rank;
+    ucc_rank_t         root      = (uint32_t)args->root;
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/reduce/reduce_knomial.c:152]: uint32_t ->ucc_rank_t

File: tools/perf/ucc_pt_coll_reduce.cc [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        ucc_mc_free(dst_header);
+    }
+    return st_src;
+exit:
+    return st_dst;
+}
+
+void ucc_pt_coll_reduce::free_coll_args(ucc_coll_args_t &args)
+{
+	bool is_root = (comm->get_rank() == args.root);
```
Comment (Sergei-Lebedev) [inline tools/perf/ucc_pt_coll_reduce.cc:64]: alignment

**Review by Sergei-Lebedev — APPROVED** [review]
pls rebase

**Comment by Sergei-Lebedev** [comment]
bot:retest

**Comment by vspetrov** [comment]
@Sergei-Lebedev looks like fails in tl_nccl_reduce after rebase on top of yours PR

**Comment by shimmybalsam** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/sbalsam/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/sbalsam/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | nnodes=8;   ppn=28; UCC_TL_UCP_REDUCE_KN_RADIX=4
-- | --
  |   |   |   |   |   |  
OSU REDUCE | UCC |   |   | coll/tuned |  
  |   |   |   |   |   |  
Size \ Latency(us) | Avg | Min | Max | Avg | Min | Max
4 | 1 | 0.36 | 8.96 | 5.29 | 4.64 | 5.93
8 | 0.85 | 0.36 | 7.52 | 6.65 | 5.99 | 7.25
16 | 0.93 | 0.36 | 7.17 | 7.7 | 6.9 | 9.18
32 | 0.99 | 0.36 | 8.12 | 5.99 | 5.32 | 6.82
64 | 0.91 | 0.37 | 8.01 | 5.75 | 5.14 | 6.37
128 | 1.29 | 0.49 | 9.08 | 7.08 | 6.34 | 7.27
256 | 1.36 | 0.49 | 11.51 | 7.38 | 6.88 | 8.07
512 | 1.54 | 0.5 | 11.93 | 7.85 | 7.23 | 8.51
1024 | 2 | 0.59 | 15.05 | 11.44 | 10.62 | 12.42
2048 | 2.28 | 0.61 | 16.14 | 13.69 | 12.8 | 14.51
4096 | 3.35 | 0.77 | 23.37 | 18.96 | 17.64 | 20.1



</body>

</html>

**Comment by shimmybalsam** [comment]
**Performance Evaluation:**

**8 nodes, 28 ppn, radix=4**

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/sbalsam/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/sbalsam/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



OSU   REDUCE | UCC |   |   | coll/tuned |   |  
-- | -- | -- | -- | -- | -- | --
  |   |   |   |   |   |   |  
Size \ Latency(us) | Avg | Min | Max | Avg | Min | Max |  
4 | 1 | 0.36 | 8.96 | 5.29 | 4.64 | 5.93 |  
8 | 0.85 | 0.36 | 7.52 | 6.65 | 5.99 | 7.25 |  
16 | 0.93 | 0.36 | 7.17 | 7.7 | 6.9 | 9.18 |  
32 | 0.99 | 0.36 | 8.12 | 5.99 | 5.32 | 6.82 |  
64 | 0.91 | 0.37 | 8.01 | 5.75 | 5.14 | 6.37 |  
128 | 1.29 | 0.49 | 9.08 | 7.08 | 6.34 | 7.27 |  
256 | 1.36 | 0.49 | 11.51 | 7.38 | 6.88 | 8.07 |  
512 | 1.54 | 0.5 | 11.93 | 7.85 | 7.23 | 8.51 |  
1024 | 2 | 0.59 | 15.05 | 11.44 | 10.62 | 12.42 |  
2048 | 2.28 | 0.61 | 16.14 | 13.69 | 12.8 | 14.51 |  
4096 | 3.35 | 0.77 | 23.37 | 18.96 | 17.64 | 20.1 |  



</body>

</html>

**Comment by lappazos** [comment]
@manjugv should we add to NEWS file that we have added support to Reduce collective?

**Comment by vspetrov** [comment]
Failing tests are fixed as part of broader change set in #293

**Pull request merged**